### PR TITLE
Allow encoding into chunks

### PR DESCRIPTION
**Pull request by chrysn**
Currently, if I have a 1200 byte CoAP buffer in a constrained device and want to send that to a UART (which is maybe buffered to a few bytes), I need to allocate a 2404-ish byte output buffer.

This PR adds a struct that can deal out pieces of data in chunks down to 2 bytes; it is a heavy refactoring of what used to be encode (which itself is now just driving that encoder through a single step).

It's a heavy PR in terms of LoC (+167-22 for a net +145, although 48 thereof is just tests), but I think it is useful for embedded applications. On the history, feel free to peruse if you feel like digging into history, but due to PR-internal refactoring, this is best reviewed as a single large diff.

### CODE REVIEW

File: src/encode.rs [hunk 1]
```diff
@@ -72,31 +71,148 @@ pub fn encode_buffered(input: Slipmux) -> Vec<u8> {
...
+
+        buffer_len_initial - buffer.len()
+    }
+
+    /// Returns true iff [`Self::encode_chunk`] would return 0.
+    #[must_use]
+    #[expect(
+        clippy::missing_const_for_fn,
+        reason = "no point in this for runtime state"
+    )]
```
Comment (Teufelchen1) [inline src/encode.rs:212]: I don't understand the reasoning. Can you elaborate?

File: src/encode.rs [hunk 2]
```diff
@@ -172,6 +288,54 @@ mod tests {
         );
     }
 
+    fn chunked<const N: usize>() {
+        extern crate alloc;
```
Comment (Teufelchen1) [inline src/encode.rs:292]: Why?

File: src/encode.rs [hunk 3]
```diff
@@ -172,6 +288,54 @@ mod tests {
...
+
+    #[test]
+    fn chunked_2() {
+        chunked::<2>();
+    }
+
+    #[test]
+    fn chunked_3() {
+        chunked::<3>();
+    }
```
Comment (Teufelchen1) [inline src/encode.rs:337]: Are you sure these functions generate / test different code? I fail to see where the generic is being used.

**Comment by Teufelchen1** [comment]
Thanks! This looks great!

File: src/encode.rs (hunk 2 above)
Comment (chrysn) [reply src/encode.rs:292]: Thanks, good catch.

I forgot the last step of the refactoring that led to this: Actually using the N. That should have been in the array length that's being pulled out.

Then, the varying chunking (2 vs 3) catches different cases of whether or not End comes at a chunk boundary, or whether the last buffer is full nor not.

File: src/encode.rs (hunk 1 above)
Comment (chrysn) [reply src/encode.rs:212]: Calling this function only makes sense after having run the encoder state machine to a point where it could be done, which is after having performed some chunk encodings (which, themselves, are not const).

Setting this const would thus not serve any practical purpose and has no benefit, but constrains future changes to implementation details to the (then) const subset.

File: src/encode.rs [hunk 4]
```diff
@@ -72,31 +71,148 @@ pub fn encode_buffered(input: Slipmux) -> Vec<u8> {
...
-        FrameType::Ip => {
-            totals += slip.encode(data, &mut buffer[totals.written..]).unwrap();
+    }
+
+    /// Writes some of the data into the output buffer.
+    ///
+    /// Returns the number of bytes written.
+    ///
+    /// Unlike many other write-style methods, this does *not* fill up the buffer to the last byte
+    /// except for the last block; this simplifies encoding escaped bytes.
```
Comment (Teufelchen1) [inline src/encode.rs:153]: But it does fill the buffer up the buffer to the last byte? Not if the buffer is very big but if the buffer is smaller than the data, it will always fill the buffer to the limit.

File: src/encode.rs (hunk 2 above)
Comment (Teufelchen1) [reply src/encode.rs:292]: Okay, but that still doesn't explain why you need to include this crate here?

I guess so that this test compiles for no_std as well as std?

Comment (chrysn) [reply src/encode.rs:292]: Ah, sorry, I thought this comment applied to the `const N`.

The thing with tests is: A `#[test]` is (at least until Cargo changes something drastically, and then there will be a migration plan) always executed with std, because of how things around it are set up. That means that the alloc (and std) crate is available, it's just that the `no_std` at the top does not automatically pull it in. So tests can't prove no-std'ness; some people try building the lib on an -unknown-unknown target in CI to show that property, but I'd say that's probably excessive here.

The alloc'd vec here is used to collect the test's data so that it can be compared. This could be done differently, eg. using the 3rd party heapless crate, or by driving comparison along the slice, but I think it'd suffer in terms of readability.

File: src/encode.rs (hunk 4 above)
Comment (chrysn) [reply src/encode.rs:153]: If it's a large buffer, it fills it up to the last byte (and there's your existing test that covers this).

Repeated small buffers will eventually be reading up the whole message.

Comment (chrysn) [reply src/encode.rs:153]: There is a corner case where if there is a large fixed-size buffer, and the application only presents that buffer to slipmux picemeal, and that buffer is odd-sized, and then in a maximal message, there's only the End byte missing, and the application only has a single byte to hadn to the library, then the "panics if <2 byte buffer" triggers even though technically things would fit.

But that'd make the usage condition more complex to describe and to test, and I don't see why anyone would do that: Either it's a fixed buffer, then it can be passed to encode_chunk as a whole, or it's some kind of streaming interface, where you get fixed buffers every time. (Or it's a streaming interface that gives you as much buffer as is free, and then it just takes a jiffy longer for a 2-long buffer to get free so that the function can be called properly again).
